### PR TITLE
SCMOD-8421: Added scaling functions and rabbit query config

### PR DESCRIPTION
**Pull request by Anthony-McGreevy**
https://portal.digitalsafe.net/browse/SCMOD-8421

This PR allows the Autoscaler to continue scaling even after the service has started to take emergency actions to protect RabbitMQ. This PR also adds configuration to the RabbitMQ memory usage querying to prevent multiple queries being issued to RabbitMQ within milliseconds of each other.
**BUILD**: http://sou-jenkins2.swinfra.net/job/Autoscaler/view/Developer/job/Autoscaler~autoscaler~anthony~CI/

### CODE REVIEW

**Comment by Anthony-McGreevy** [comment]
@dermot-hardy Just in case we decide it is worth doing a quick fix and patch release for.

File: autoscale-workload-rabbit/README.md [hunk 1]
```diff
@@ -25,7 +25,9 @@
  - vhost: the RabbitMQ vhost that contains the queues
  - profiles: map of profile name to `RabbitWorkloadProfile` objects, which
   represent different scaling profiles for this implementation (see below).
-  Note there *must* be a profile named "default"
+  Note there *must* be a profile named "default"  
+ - memoryQueryRequestFrequency: The number of whole minutes that the service should wait between issuing requests to check rabbitmq's
```
Comment (dermot-hardy) [inline autoscale-workload-rabbit/README.md]: Change to seconds.

File: autoscale-marathon-container/src/main/config/cfg~caf~autoscaler~RabbitWorkloadAnalyserConfiguration.js [hunk 2]
```diff
@@ -18,6 +18,7 @@
             || ("http://" + (getenv("CAF_RABBITMQ_HOST") || "rabbitmq") + ":" + (getenv("CAF_RABBITMQ_MGMT_PORT") || "15672")),
     rabbitManagementUser: getenv("CAF_RABBITMQ_MGMT_USERNAME") || getenv("CAF_RABBITMQ_USERNAME") || "guest",
     rabbitManagementPassword: getenv("CAF_RABBITMQ_MGMT_PASSWORD") || getenv("CAF_RABBITMQ_PASSWORD") || "guest",
+    memoryQueryRequestFrequency: getenv("CAF_AUTOSCALER_RABBITMQ_MEMORY_QUERY_FREQ") || 1,
```
Comment (dermot-hardy) [inline autoscale-marathon-container/src/main/config/cfg~caf~autoscaler~RabbitWorkloadAnalyserConfiguration.js]: Change default to 10 seconds.

File: autoscale-workload-rabbit/src/main/java/com/hpe/caf/autoscale/workload/rabbit/RabbitSystemResourceMonitor.java [hunk 3]
```diff
@@ -24,6 +24,8 @@
 import java.util.Base64;
 import java.util.Iterator;
 import java.util.concurrent.TimeUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
```
Comment (dermot-hardy) [inline autoscale-workload-rabbit/src/main/java/com/hpe/caf/autoscale/workload/rabbit/RabbitSystemResourceMonitor.java]: Unused

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 4]
```diff
@@ -126,9 +128,9 @@ private void handleAnalysis()
                 LOG.debug("Scaling checks for service {}", serviceRef);
                 action = analyser.analyseWorkload(instances);
             }
-
-            action = governor.govern(serviceRef, action);
-
+            if(shutdownPriority == -1 || shouldNotGoven(currentMemoryLoad, shutdownPriority)){
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: Should this be `shouldGovern`?

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 5]
```diff
@@ -257,4 +255,29 @@ private void handleAlerterDispatch(final double memLoad) throws ScalerException
             alertDispatcher.dispatchAlert(emailBody);
         }
     }
+    
+    private int establishMemLimitReached(final double currentMemoryLoad)
+    {
+        if (currentMemoryLoad >= resourceConfig.getResourceLimitOne()) {
+            return 1;
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: Function faulty

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java (hunk 5 above)
Comment (Anthony-McGreevy) [reply autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java]: Fixed logic

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java [hunk 6]
```diff
@@ -83,6 +95,10 @@ public ScalingAction govern(String serviceRef, ScalingAction action) {
                     if (lastInstanceInfo.getTotalInstances() < scalingConfiguration.getMinInstances()) {
                         return new ScalingAction(ScalingOperation.SCALE_UP,
                                                  scalingConfiguration.getMinInstances() - lastInstanceInfo.getTotalInstances());
+                    } else if (!shouldGovern(currentMemoryLimitStage, shutdownPriority)) {
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java]: - Terminology is confusing (you're in a method called `govern` introducing a function called `shouldGovern`)
- Branch is duplicated (reorder to eliminate)
- I don't see how this decision can be taken without reference to why are other service's minimum instances aren't met.

**Review by dermot-hardy — CHANGES_REQUESTED** [review]

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java [hunk 7]
```diff
@@ -209,4 +205,18 @@ private void handleAlerterDispatch(final double memLoad) throws ScalerException
             alertDispatcher.dispatchAlert(emailBody);
         }
     }
+    
+    private int establishMemLimitReached(final double currentMemoryLoad)
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/ScalerThread.java:209]: 3, 2, 1, 0

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java [hunk 8]
```diff
@@ -133,10 +144,37 @@ private boolean otherServicesMinimumInstancesMet(String serviceRef){
...
+                if(isScaledDownForMemoryManagement(currentMemoryLimitStage, lastInstanceInfo.getShutdownPriority())){
+                   continue;
+                }
                 return false;
             }
         }
         return true;
     }
+
+    private boolean isScaledDownForMemoryManagement(final int currentMemoryLoadLimit, final int shutdownPriority)
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java]: - `static`
- Consider renaming to `shouldBeScaledDown`.

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java [hunk 9]
```diff
@@ -133,10 +144,37 @@ private boolean otherServicesMinimumInstancesMet(String serviceRef){
...
+                return shutdownPriority <= stageOneShutdownPriorityLimit;
+            }
+            case 2: {
+                return shutdownPriority <= stageTwoShutdownPriorityLimit;
+            }
+            case 3: {
+                return shutdownPriority <= stageThreeShutdownPriorityLimit;
+            }
+            default: {
+                return true;
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java]: Should be `return false;`.

File: autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java [hunk 10]
```diff
@@ -89,7 +101,6 @@ public ScalingAction govern(String serviceRef, ScalingAction action) {
                         //Gradually reduce the totalInstances by a percentage until Minimums are met.
                         //This should be configurable, however there should be a Governor specific configuration
                         //to allow different Governor implementations to be added without polluting the AutoscaleConfiguration
-
```
Comment (dermot-hardy) [inline autoscale-core/src/main/java/com/hpe/caf/autoscale/core/GovernorImpl.java]: Add line back in

**Review by dermot-hardy — APPROVED** [review]
Ensure you definitely fix that return value one.

**Pull request merged**